### PR TITLE
Flatten nested optionals resulting from try? and optional sub-expressions

### PR DESCRIPTION
**Pull request by bjhomer**
This PR implements the flattening of a double-optional type produced from code like `let x = try? optionalThing()`. The pitch thread can be found here: https://forums.swift.org/t/make-try-optional-chain-flattening-work-together/7415/55

### CODE REVIEW

**Comment by bjhomer** [comment]
@brentdax Rebased on master, tests are passing locally. Could you run the compatibility suite tests?

**Comment by brentdax** [comment]
@swift-ci please test source compatibility

**Review by rudkx — CHANGES_REQUESTED** [review]
This change would benefit from additional test cases.

For example, what happens with something like:

```
func foo() throws -> Int? { return 1 }
let x = (try? foo())!!
```

With the new semantics I would expect this to fail type checking, hopefully with a good diagnostic.

It's possible we already have some interesting test cases along these lines, but they are in files not currently compiled with `-swift-version 5`. Can you take a look and add some more comprehensive tests to a file that is compiled with `-swift-version 5`?

**Comment by bjhomer** [comment]
Yep, I'll get on this.

**Review by xedin — COMMENTED** [review]
I agree with @rudkx, it would be great to add more tests for this, otherwise changes to Sema LGTM!

**Comment by bjhomer** [comment]
I have added a bunch more tests. A couple of them are currently failing; it appears I need to make some changes for cases where we used to suggest changing `try?` to `try!`.

**Comment by bjhomer** [comment]
@xedin, @rudkx Tests have been added. I discovered that a fix-it would suggest changing `try?` into `try!` in cases where that would no longer fix anything, so that has also been addressed. I think we're ready for further review.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

File: test/Parse/try.swift [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-typecheck-verify-swift
+// RUN: %target-typecheck-verify-swift -swift-version 5
```
Comment (rudkx) [inline test/Parse/try.swift]: Can you split this into two tests, one for ambient-version mode (with no changes to the existing test) and one for `-swift-version 5`?

We still want to test the older behavior as long as it is supported.

**Comment by rudkx** [comment]
@nkcsgexi I cannot seem to currently add you as a reviewer. Can you take a look at the migrator changes?

**Comment by nkcsgexi** [comment]
hmm, interesting. I've added myself as a reviewer and will take a look.

**Review by rudkx — CHANGES_REQUESTED** [review]
I looked at the AST & Sema changes as well as the tests. LGTM. Once you split out the new tests into a separate file for `-swift-version 5` I think that portion is good to go.

File: test/Parse/try.swift (hunk 1 above)
Comment (bjhomer) [reply test/Parse/try.swift]: Yeah, I can do that.

**Comment by bjhomer** [comment]
Regarding the migration, I have a test file that I've been using locally. I would be happy to add it as an automated test, if I could figure out how. Is there some example of how to run a test like that in the repository right now?

**Comment by rudkx** [comment]
@nkcsgexi Can you answer the question re: migration testing?

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@bjhomer you can follow some examples in test/migrator. Taking `remove_override.swift` as an example, it's both test driver and the code example before migration. After running the migrator, the result should be identical to `remove_override.swift.expected`.

**Review by rudkx — APPROVED** [review]

**Comment by bjhomer** [comment]
@nkcsgexi: Thanks for that example. I'm bulking up these tests, and it's pointing out a few things I need to improve.

**Comment by bjhomer** [comment]
The migration tests are ready to go. I've simplified the approach a bit after discovering some problems with the earlier one.

File: test/Parse/try_swift4.swift [hunk 2]
```diff
@@ -0,0 +1,260 @@
+// RUN: %target-typecheck-verify-swift -swift-version 4
```
Comment (bjhomer) [inline test/Parse/try_swift4.swift]: Even though git is showing this as a new file, this just the existing `try.swift` test with a few new tests added to differentiate between the swift4 and the swift 5 behavior..

**Comment by tkremenek** [comment]
This would be good to land before Swift 5 branches one last time from `master` on November 16.

**Comment by bjhomer** [comment]
Agreed. I think everything is ready to go, but if there's any further feedback, I'll be responsive.

**Review by nkcsgexi — APPROVED** [review]

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@nkcsgexi @rudkx Okay to merge?

**Comment by nkcsgexi** [comment]
The migrator part LGTM to merge!

**Comment by rudkx** [comment]
Yes, merging.

**Comment by rudkx** [comment]
@bjhomer It looks like this broke the foundation build: https://ci.swift.org/job/swift-PR-Linux-smoke-test/11232/consoleFull#-1189064102373bf607-d277-47c6-812e-27acc852624c

```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift-corelibs-foundation/TestFoundation/TestURL.swift:38:38: error: cannot force unwrap value of non-optional type '[String : Any]'
13:46:22     guard let parsingTests = testRoot![kURLTestParsingTestsKey] as? [Any] else {
```

**Comment by rudkx** [comment]
Okay, I don't see how this passed PR testing yesterday but would now fail today. I might be missing something that changed in foundation. Reverting for now, though.

**Comment by bjhomer** [comment]
Fixed in https://github.com/apple/swift-corelibs-foundation/pull/1777

**Pull request merged**